src/main/java/com/minecolonies/coremod/colony/buildings/modules/AbstractCraftingBuildingModule.java

```diff
@@ -591,16 +591,20 @@ public void improveRecipe(IRecipeStorage recipe, int count, ICitizenData citizen
                   recipe.getPrimaryOutput(),
                   Blocks.AIR);
 
-                replaceRecipe(recipe.getToken(), IColonyManager.getInstance().getRecipeManager().checkOrAddRecipe(storage));
-
-                // Expected parameters for RECIPE_IMPROVED are Job, Result, Ingredient, Citizen
-                Component jobComponent = MessageUtils.format(citizen.getJob().getJobRegistryEntry().getTranslationKey()).create();
-                MessageUtils.format(RECIPE_IMPROVED + citizen.getRandom().nextInt(3),
-                    jobComponent,
-                    recipe.getPrimaryOutput().getHoverName(),
-                    reducedItem.getItemStack().getHoverName(),
-                    citizen.getName()
-                ).sendTo(building.getColony()).forAllPlayers();
+                final IToken<?> token = IColonyManager.getInstance().getRecipeManager().checkOrAddRecipe(storage);
+                if (isRecipeCompatibleWithCraftingModule(token))
+                {
+                    replaceRecipe(recipe.getToken(), IColonyManager.getInstance().getRecipeManager().checkOrAddRecipe(storage));
```

Comment:
Shouldn't this just use `token` rather than calling `checkOrAddRecipe` again?  (Current will behave correctly but is inefficient.)